### DISCUSSION TITLE
Installing Pytorch3d without Cuda

### DISCUSSION THREAD

**Discussion by YanivHollander** [facebookresearch/pytorch3d#1117] [General]
Hi,

I am trying to install Pytorch3D on a Ubuntu 20.04 LTS which happens to not have GPU. Is it possible to install Pytorch3d for CPU only runs (for debugging purposes, e.g.)? If so, what is the workflow I should take to install this package in a Python 3.8.10 based virtual environment?

**Comment by bottler** [comment]
Yes, this works. You need to install from source. Start with an environment which has torch and torchvision. I recommend a conda environment. (You can do this with `conda install -c pytorch pytorch=1.9.1 torchvision cpuonly` or following  https://pytorch.org/get-started .  

* Verify that `import torch; print(torch.cuda.is_enabled())` works in python and gives `False`.

* Then install fvcore and iopath with `pip install fvcore iopath` or `conda install -c fvcore -c iopath -c conda-forge fvcore iopath`.

* Make sure you have compilers (e.g. is there a `g++` command available). If not, `sudo apt install build-essential` might be enough.

* Then install pytorch3d from source with pip install "git+https://github.com/facebookresearch/pytorch3d.git@stable" or by cloning this repository, `git checkout stable` and `pip install -e .`

**Reply by YanivHollander** [reply]
Thank you, this workflow worked! I am now able to import `pytorch3d.renderer` without a CUDA support. Just a typo correction for reference: `torch.cuda.is_available()`, instead of what you wrote.

**Reply by Joyyy821** [reply]
Hi!
I am also installing Pytorch3d without Cuda and came across this discussion. I have followed the steps, but my terminal crashed without any error message while executing the last step `pip install -e .`
I monitored the memory usage using `htop`, it seems like the installation used all of my 16G RAM and an extra 16G swap (I have 32G swap in total, but the terminal just closed itself when there was 16G swap being used). Is this normal and should I add more swap? 

I have successfully installed pytorch3d before on ubuntu 18.04 (with python 3.6 if I remember correctly). I recently upgraded to ubuntu 22.04 and try to install pytorch3d on python 3.10.12. My pytorch version is 2.1.0.

**Reply by bottler** [reply]
@Joyyy821 can you reduce the parallelism of the build. E.g. set MAX_JOBS=1

**Reply by Joyyy821** [reply]
That just solved my problem perfectly! Thank you so much!!

**Reply by benmagyar03** [reply]
Hi, I am also trying to install CPU only PyTorch3D on linux, in a virtual environment with python 3.10.14. I am not very familiar with Linux, but I have followed all the steps you outlined above and in the pytorch3d github. 

When I try to run pip install "git+https://github.com/facebookresearch/pytorch3d.git", I always get an error: "ModuleNotFoundError: No module named 'torch' ". 

However, when I run "pip list | grep torch" I have torch 2.5.1+cpu, torchaudio 2.5.1+cpu, and torchvision 0.20.1+cpu installed. Do you have any ideas for how to fix this error? I would really appreciate any help

**Reply by bottler** [reply]
@benmagyar03 You need to make sure you can import torch in python before you install PyTorch3D if you can't that's not really a quite 3-D problem. If you need any more help, you could open an issue or, even better, ask somewhere not pytorch3d specific.
